narwhal/types/src/primary.rs

```diff
@@ -166,6 +243,90 @@ impl BatchV1 {
     }
 }
 
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary)]
+pub struct BatchV2 {
+    pub transactions: Vec<Transaction>,
+    pub versioned_metadata: VersionedMetadata,
+    pub epoch: Epoch,
+}
+
+impl BatchAPI for BatchV2 {
+    fn transactions(&self) -> &Vec<Transaction> {
+        &self.transactions
+    }
+
+    fn transactions_mut(&mut self) -> &mut Vec<Transaction> {
+        &mut self.transactions
+    }
+
+    fn metadata(&self) -> &Metadata {
+        unimplemented!("BatchV2 does not have a Metadata field");
+    }
+
+    fn metadata_mut(&mut self) -> &mut Metadata {
+        unimplemented!("BatchV2 does not have a Metadata field");
+    }
+
+    fn versioned_metadata(&self) -> &VersionedMetadata {
+        &self.versioned_metadata
+    }
+
+    fn versioned_metadata_mut(&mut self) -> &mut VersionedMetadata {
+        &mut self.versioned_metadata
+    }
+
+    fn epoch(&self) -> &Epoch {
+        &self.epoch
+    }
+}
+
+impl BatchV2 {
+    pub fn new(
+        transactions: Vec<Transaction>,
+        protocol_config: &ProtocolConfig,
+        epoch: Epoch,
+    ) -> Self {
+        Self {
+            transactions,
+            versioned_metadata: VersionedMetadata::new(protocol_config),
+            epoch,
+        }
+    }
+
+    pub fn size(&self) -> usize {
+        self.transactions.iter().map(|t| t.len()).sum()
+    }
+}
+
+// TODO: Remove once we have upgraded to protocol version 12.
```

Comment:
I was thinking about this and I was wondering if after we upgrade to v12 we can then remove BatchV1. In which case IIUC no one can send it to us because we don't have that type in the enum?